dds/DCPS/DataReaderImpl.cpp

```diff
@@ -3632,6 +3576,49 @@ void DataReaderImpl::OnDataAvailable::execute()
     subscriber->set_status_changed_flag(::DDS::DATA_ON_READERS_STATUS, false);
   }
 }
+void DataReaderImpl::initialize_lookup_maps()
+{
+  // These all start at 1 (0 mask is bogus) and include the full mask (any)
+  for (CORBA::ULong is = 1; is <= MAX_SAMPLE_STATE_MASK; ++is) {
+    for (CORBA::ULong iv = 1; iv <= MAX_VIEW_STATE_MASK; ++iv) {
+      for (CORBA::ULong ii = 1; ii <= MAX_INSTANCE_STATE_MASK; ++ii) {
+        combined_state_lookup_[to_combined_states(is, iv, ii)] = HandleSet();
+      }
+    }
+  }
+  // catch-all for "bogus" lookups
+  combined_state_lookup_[0] = HandleSet();
+}
+
+void DataReaderImpl::update_lookup_maps(const SubscriptionInstanceMapType::iterator& input)
+{
+  for (LookupMap::iterator it = combined_state_lookup_.begin(); it != combined_state_lookup_.end(); ++it) {
+    if (it->first == 0) continue;
+    CORBA::ULong sample_states, view_states, instance_states;
+    split_combined_states(it->first, sample_states, view_states, instance_states);
+    if (input->second->matches(sample_states, view_states, instance_states)) {
+      it->second.insert(input->first);
+    } else {
+      it->second.erase(input->first);
+    }
+  }
```

Comment:
While this is "constant time", it should be not depend on the length of `combined_state_lookup_`.